File: Problems/823_Binary_Trees_With_Factors.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <math.h>

using namespace std;
class Solution {
public:
    int numFactoredBinaryTrees(vector<int>& arr) {
        sort(arr.begin(), arr.end());
        unordered_map<int, long long> dp;
        long long total = 0;
        long long mod = pow(10, 9) + 7;
        for (int i = 0; i < arr.size(); ++i) {
            int father = arr[i];
            dp[father] = 1;
            for (int j = 0; j < i; ++j) {
                int child = arr[j];
                if (child > sqrt(father)) break;
                if (father % child || dp.count(father/child) == 0) continue;
                long long tmp;
                if (child == father / child){
                    tmp = dp[child] * dp[child] % mod;
                }else{
                    tmp = dp[child] * dp[father/child] * 2 % mod;
                }
                tmp += dp[father];
                tmp %= mod;
                dp[father] = tmp;
            }
            total += dp[father];
            total %= mod;
        }
        return (int)total;
    }
};

File: Problems/823_Binary_Trees_With_Factors.py
from typing import List

class Solution:
    def numFactoredBinaryTrees(self, arr: List[int]) -> int:
        arr.sort()
        mod = 10 ** 9 + 7
        dp = {}
        total = 0
        for i, father in enumerate(arr):
            dp[father] = 1
            for j in range(i):
                child = arr[j]
                # we don't have to check the child bigger than the square root of father
                if child > father **0.5:
                    break
                # not divisible or not inside the dp
                if father % child or father // child not in dp:
                    continue
                if child == father // child:
                    dp[father] += dp[child] * dp[child]
                else:
                    dp[father] += dp[child] * dp[father // child] * 2
            dp[father] %= mod
            total += dp[father]
            total %= mod
        return total

if __name__ == '__main__':
    sol = Solution()
    arr = [2,4]
    print(sol.numFactoredBinaryTrees(arr))